docs: native histogram migration alternatives (#8864)
grafana/mimir: docs/sources/mimir/send/native-histograms/_index.md

Use the latest version of [Grafana Alloy](https://grafana.com/docs/alloy/<ALLOY_VERSION>).

> 1. To scrape native histograms, you need to set the `scrape_protocols` argument in the `prometheus.scrape` component to specify `PrometheusProto` as the first protocol to negotiate and `scrape_classic_histograms = true` in order to scrape both classic and native histograms.

Author: I'd rather keep it in one step, since it's easier for people to copy paste once.

   ```
    scrape_protocols = ["PrometheusProto", "OpenMetricsText1.0.0", "OpenMetricsText0.0.1", "PrometheusText0.0.4"]
    scrape_classic_histograms = true
   ```

   For more information, refer to [prometheus.scrape](https://grafana.com/docs/alloy/<ALLOY_VERSION>/reference/components/prometheus/prometheus.scrape/) in the Grafana Alloy documentation.

   {{< admonition type="note" >}}
   In certain situations, the protobuf parsing changes the number formatting of
   the `le` labels of conventional histograms and the `quantile` labels of
   summaries. Typically, this happens if the scraped target is instrumented with
   [client_golang](https://github.com/prometheus/client_golang) provided that
   [promhttp.HandlerOpts.EnableOpenMetrics](https://pkg.go.dev/github.com/prometheus/client_golang/prometheus/promhttp#HandlerOpts)
   is set to `false`. In such a case, integer label values are represented in the
   text format as such, e.g. `quantile="1"` or `le="2"`. However, the protobuf parsing

>    changes the representation to float-like (following the OpenMetrics

Author: I've changed it to talk about the [fractional](https://en.wikipedia.org/wiki/Decimal) part of decimal numbers. That should be super precise.

   specification), so the examples above become `quantile="1.0"` and `le="2.0"` after
   ingestion into Prometheus, which changes the identity of the metric compared to

>    what was ingested before via the text format.

Author: applied

          Buckets: prometheus.DefBuckets,  // If buckets weren't already defined.
          NativeHistogramBucketFactor: 1.1,
          NativeHistogramMaxBucketNumber: 100,
          NativeHistogramMinResetDuration: 1*time.Hour,
   })
   ```

   ```java
   static final Histogram requestLatency = Histogram.build()
      .name("requests_latency_seconds")
      .help("Histogram of request latency in seconds")
      .classicUpperBounds(Histogram.Builder.DEFAULT_CLASSIC_UPPER_BOUNDS)  // If upper bounds weren't already defined.
      .nativeInitialSchema(3)
      .nativeMaxNumberOfBuckets(100)
      .nativeResetDuration(1, TimeUnit.HOURS)
      .register();
   ```

   {{< /code >}}

1. Let Prometheus or Grafana Alloy scrape both classic and native histograms for metrics that have both defined.
1. Send native histograms to remote write, along with the existing classic histograms.
1. Modify dashboards to use the native histogram metrics. Refer to [Visualize native histograms](https://grafana.com/docs/mimir/<MIMIR_VERSION>/visualize/native-histograms/) for more information.

   Use one of the following strategies to update dashbaords.

>    - (Recommended) Add new dashboards with the new native histogram queries. This solution requires looking at different dashboards for data before and after the migration, until data before the migration is removed due to passing its retention time. Thus it is recommended to publish the new dashboard when sufficient time has passed to serve users with the new data.

Reviewer: ```suggestion
   - (Recommended) Add new dashboards with the new native histogram queries. This solution requires looking at different dashboards for data before and after the migration, until data before the migration is removed due to passing its retention time. You can publish the new dashboard when sufficient time has passed to serve users with the new data.
```

Reviewer: I think we can simplify this sentence a bit.

>    - Add a dashboard variable to your dashboard to enable switching between classic histograms and native histograms. There isn't support for selectively enabling and disabling queries in Grafana yet ([issue 79848](https://github.com/grafana/grafana/issues/79848)). As a workaround, add the dashboard variable `latency_metrics`, for example, and assign it a value of either `-1` or `1`. Then, add the following two queries to the panel:

Reviewer: ```suggestion
   - Add a dashboard variable to your dashboard to enable switching between classic histograms and native histograms. There isn't support for selectively enabling and disabling queries in Grafana ([issue 79848](https://github.com/grafana/grafana/issues/79848)). As a workaround, add the dashboard variable `latency_metrics`, for example, and assign it a value of either `-1` or `1`. Then, add the following two queries to the panel:
```

Reviewer: We try to keep docs always current, without referencing support that may be added in the future.

          Buckets: prometheus.DefBuckets,  // If buckets weren't already defined.
          NativeHistogramBucketFactor: 1.1,
          NativeHistogramMaxBucketNumber: 100,
          NativeHistogramMinResetDuration: 1*time.Hour,
   })
   ```

   ```java
   static final Histogram requestLatency = Histogram.build()
      .name("requests_latency_seconds")
      .help("Histogram of request latency in seconds")
      .classicUpperBounds(Histogram.Builder.DEFAULT_CLASSIC_UPPER_BOUNDS)  // If upper bounds weren't already defined.
      .nativeInitialSchema(3)
      .nativeMaxNumberOfBuckets(100)
      .nativeResetDuration(1, TimeUnit.HOURS)
      .register();
   ```

   {{< /code >}}

1. Let Prometheus or Grafana Alloy scrape both classic and native histograms for metrics that have both defined.
1. Send native histograms to remote write, along the existing classic histograms.
1. You can now start modifying dashboards to use the new native histogram metrics and get familiar with the new query syntax, see [Visualize native histograms](https://grafana.com/docs/mimir/<MIMIR_VERSION>/visualize/native-histograms/).

   There are different strategies to updating the dashboards:

>    1. Add new dashboards with the new native histogram queries. This is the cleanest and recommended solution, but does require looking at different dashboards for data before and after the migration, while the retention time of the data runs out. Thus it is recommended to publish the new dashboard when sufficient time has passed to serve users with the new data.

Author: reworded

>    1. Add a dashboard variable on top of your dashboard to be able to switch between classic histograms and native histograms. Currently there's no direct support for selectively enabling and disabling queries in Grafana ([issue 79848](https://github.com/grafana/grafana/issues/79848)), however there's a relatively easy workaround. Let the variable take the values -1 or 1 and call it for example `latency_metrics`. Add the following two queries to the panel:

Author: made an update, plz check again

      ```
      <classic_query> < ($latency_metrics * +Inf)
      ```

      and separately

      ```
      <native_query> < ($latency_metrics * -Inf)
      ```

      Where `classic_query` is the original query and `native_query` is the same but using native histogram query syntax. This technique is employed for Mimir's own dashboards, see for example the [Overview dashboard](https://github.com/grafana/mimir/blob/main/operations/mimir-mixin-compiled/dashboards/mimir-overview.json).

      This solution allows users to switch between the classic histogram and the native histogram without going to a different dashboard.

   1. Replace the existing classic queries with a modified query:

      ```
      <classic_query>
      ```

      replaced with

      ```
      <native_query> or <classic_query>
      ```

      Where `classic_query` is the original query and `native_query` is the same but using native histogram query syntax.

      {{< admonition type="warning" >}}

>       Using the PromQL operator `or` can lead to unexpected results. For example if the query uses a range of 7 days, such as `sum(rate(http_request_duration_seconds[7d]))` then this query will return a value as soon as there are two native histograms samples present before the end time of the query, thus the 7 day rate will be calculated from a couple of minutes worth of data and not 7 days. This means that the result will be very inaccurate around the time when native histograms were started to be scraped and this inaccuracy will always be there when looking at the graph for that time.

Reviewer: ```suggestion
      Using the PromQL operator `or` can lead to unexpected results. For example, if a query uses a range of seven days, such as `sum(rate(http_request_duration_seconds[7d]))`, then this query returns a value as soon as there are two native histograms samples present before the end time specified in the query. In this case, the seven day rate is calculated from a couple of minutes, rather than seven days, worth of data. This results in an inaccuracy in the graph around the time you started scraping native histograms.
```

      {{< /admonition >}}

1. Start adding _new_ recording rules and alerts to use native histograms. Do not remove the old recording rules and alerts at this time.

> 1. It is important to keep scraping both classic and native histograms for as long as the longest range in the recording rules and alerts, plus a day. This is the minimum, but it is recommended to keep scraping both until the new rules and alerts can be verified.

Reviewer: ```suggestion
1. It is important to keep scraping both classic and native histograms for at least the period of the longest range in your recording rules and alerts, plus one day. This is the minimum amount of time, but it's recommended to keep scraping both data types until the new rules and alerts can be verified.
```

>    Here's an example why: if you have an alert that calculates the rate of requests, for example `sum(rate(http_request_duration_seconds[7d]))`, this query will look at the data from the last 7 days plus the Prometheus [lookback period](https://prometheus.io/docs/prometheus/latest/querying/basics/#staleness). When you start sending native histograms, the data will not be there for the whole 7 days, thus the results maybe unreliable for alerting.

Reviewer: ```suggestion
   For example, if you have an alert that calculates the rate of requests, such as `sum(rate(http_request_duration_seconds[7d]))`, this query looks at the data from the last seven days plus the Prometheus [lookback period](https://prometheus.io/docs/prometheus/latest/querying/basics/#staleness). When you start sending native histograms, the data isn't there for the entire seven days, and therefore, the results might be unreliable for alerting.
```

1. After configuring native histogram collection, choose one of the following ways to stop collecting classic histograms.

>    - Remove the custom bucket definition, `Buckets`/`classicUpperBounds`, from the instrumentation. In case of Java, also use the option `nativeOnly()`. See the examples in [Instrument application with Prometheus client libraries](#instrument-application-with-prometheus-client-libraries) above.

Reviewer: ```suggestion
   - Remove the custom bucket definition, `Buckets`/`classicUpperBounds`, from the instrumentation. In Java, also use the `nativeOnly()` option. Refer to the examples in [Instrument application with Prometheus client libraries](#instrument-application-with-prometheus-client-libraries).
```
